Add ability to toggle visibility of graph series in explore section

miqh opened the pull request:
Hello,

This pull request attempts to address #13522.

The implemented UX for toggling series mirrors the behaviour seen for the dashboard graph plugin:

* Single clicking a series label isolates visibility to that series
* Single clicking an isolated series label restores visibility of all series
* Single clicking a series label with a modifier key held (i.e. <kbd>ctrl</kbd>, <kbd>shift</kbd>, <kbd>meta</kbd>) inverts the visibility inclusion

Also noticed some incomplete (?) groundwork being laid for user options, presumably to support user preferences such as timezone setting. Hopefully these changes don't get in the way too much.

CC: @davkal

torkelo commented on public/app/features/explore/Graph.tsx:
```diff
@@ -143,52 +174,64 @@ export class Graph extends PureComponent<GraphProps, GraphState> {
+          hiddenSeries: nextHiddenSeries,
+        };
+      },
+      () => this.draw(false)
```
wont set state cause a componentDidUpdate & draw already?

davkal replied:
In the current state, `draw` would not be called because componentDidUpdate only calls it when certain things changed, and `hiddenSeries` it not among them.

It's a valid point though, to gate the execution of `draw` through componentDidUpdate like it is for other props/state items. I'd add a check for `hiddenSeries` there and get rid of the `updateDynamicOptions` logic.

miqh replied:
Ah, thanks for the advice on `componentDidUpdate`—my rookie React skills definitely showing here.

My original intention of having `updateDynamicOptions` is due to the `range` prop that's fed into the `Graph` component. It seems to constantly update itself. This meant toggling series also undesirably advanced the time range for the x-axis.

Any advice here? It's like I need to cache the last `dynamicOptions` that came about from running a query.

torkelo commented on public/app/features/explore/Graph.tsx:
```diff
@@ -143,52 +174,64 @@ export class Graph extends PureComponent<GraphProps, GraphState> {
+      (state, props) => {
+        const { data } = props;
+        const { hiddenSeries } = state;
+        const hidden = hiddenSeries.has(series);
```
not sure how well this set will work, will it not be using reference equality? So next refresh, with new TimeSeries instances it won't match anymore?

davkal commented:
Works great until timeseries are being changed. Maintaining hidden state of series between updates is very desirable, but makes the state management a lot more tricky.
The Graph snapshots need to be updated too.

miqh commented:
@davkal, good point about maintaining series visibility state across queries. Any pointers on how to tackle this? I think one difficulty here is no guaranteed UID property on returned `TimeSeries` objects. For example, while unusual, you could have the exact same query on two different `QueryRow` components, and you'd like for one to be off and other on.

Also, what did you mean by graph snapshots? Maybe I'm looking in the wrong areas, but I can't seem to find a feature that takes snapshots of series.

davkal commented:
@miqh yes, the dashboard graph panel faces the same issue, and it currently uses the series alias as the hidden identifier. I'm happy to accept a PR using that.

A side note: The dashboard graph legend (app/plugins/panel/graph/legend) was rewritten in React last week, so perhaps it's worth reusing it instead of the makeshift Legend in the explore directories. But could also be done separately.

davkal commented:
The snapshots I was referring to are from jest: https://jestjs.io/docs/en/snapshot-testing
The Graph.tsx is covered by one of those tests, and this PR changed the HTML markup of a graph. If you think the markup change is ok you can run the test, and then say "update snapshots" after they have been reported as mismatch. Then new snapshots files are written to disk, which you should then commit as part of this PR.

miqh committed:
Add visibility toggle for explore graph series

The implemented toggling UX is similar to how the dashboard graph plugin
behaves. Also incorporates review feedback to persist series visibility
state by means of the alias property, with the limitation it carries
too.

Related: #13522

miqh commented:
@davkal, thanks for the hint about Jest snapshots. Funky feature, TIL item for me! :slightly_smiling_face:

So just to summarise some of the other changes arising out of review:

* Series visibility state should now be persisted between queries, tracked by the `alias` property, mirroring the dashboard graph at present.
* Pruning hidden series elements no longer part of those available as of the latest query
* Deferring the constantly updating graph range to be addressed in another pull request

Hopefully I haven't missed anything else.

davkal commented:
That looks like a good plan. (Plus the necessary snapshot updates)

miqh commented:
@davkal, sorry for the delay—was there any other outstanding changes you were waiting on for merge?

davkal commented on public/app/features/explore/utils/set.ts:
```diff
@@ -0,0 +1,34 @@
+  while (true) {
+    const { value, done } = it.next();
+    if (b.has(value)) {
+      return false;
```
Can we have a test for util functions? I'm tripping over this line.

miqh replied:
Ended up being good catch—tests revealed I'd implemented the operations incorrectly.

davkal commented:
Behavior looks good now, although the visibility is maintained across query changes (run query A, toggle visibility, then run query B, then run query A again). I'm a bit undecided if that's a bug or a feature. At the same time I'm at a loss of coming up with a signal that indicates query change other than the query itself. But we need to support query objects, not just strings, in the future and they may not be object identical even though the query is the same. Hence, I'd leave this as is. I think this can go in when the sets tests have been added.

miqh committed:
Fix set utilities for explore section

Discovered implementation issues with the set utilities used in the
explore section after adding some tests. Added tests to reinforce the
expected behaviour of the utility functions.

Related: #13522

miqh commented:
@davkal, that's another good point about the hidden series persistence.

I guess we might be coming full circle back to how I had originally tracked hidden state via the `TimeSeries` object itself, but didn't have persistence between query runs.

Another counter-intuitive (IMHO) case: run query A, toggle some series, run query B, toggle some series, run query A again, observe state reset. As long as no toggling happens after query B and query A is run again, the state is persisted.

I tried hooking into `componentDidUpdate()` to see if I could update the `hiddenSeries` state based off the `data` prop (smells of an anti-pattern regardless). Every time queries ran and had changed, I was hoping to use the hook to wipe out any stale `hiddenSeries` values. However, it was naive because the `data` prop gets blown away when the query is in progress, which causes all `hiddenSeries` values to be trashed.

Not sure at this point either.

davkal commented:
I think the persistence across query changes is ok. With the set logic fixed, I think this can go in.
